integration/service/ranking/trafficapplyingmodel/preferlivetraffic/model.go

```diff
@@ -0,0 +1,197 @@
+	}
+	if m.LiveTrafficQuerier != nil && enableLiveTraffic { // requires live traffic data
+		liveTrafficSourceNameIndex = len(l.Annotation.Metadata.DataSourceNames)
+		l.Annotation.Metadata.DataSourceNames = append(l.Annotation.Metadata.DataSourceNames, trafficapplyingmodel.SourceNameLiveTraffic)
+	}
+
+	// recalc
+	var sumOriginalAnnotationDuration, sumOriginalAnnotationDistance, sumOriginalAnnotationWeight float64
+	var newLegDuration, newLegWeight float64
+	var appliedHistoricalSpeedCount, appliedLiveTrafficSpeedCount int
+	for i := 0; i < waysCount; i++ {
+		wayID := l.Annotation.Ways[i]
+
+		sumOriginalAnnotationDistance += l.Annotation.Distance[i]
+		sumOriginalAnnotationDuration += l.Annotation.Duration[i]
+		sumOriginalAnnotationWeight += l.Annotation.Weight[i]
+
+		if m.HistoricalSpeedQuerier != nil && enableHistoricalSpeed {
+			if !trafficapplyingmodel.IsInvalidSpeed(l.Annotation.HistoricalSpeed[i]) { // valid historical speed
+				l.Annotation.DataSources[i] = historicalSpeedSourceNameIndex
+				l.Annotation.Speed[i] = l.Annotation.HistoricalSpeed[i]
+				l.Annotation.Duration[i] = l.Annotation.Distance[i] / l.Annotation.Speed[i] // not include turn duration
+				l.Annotation.Weight[i] = l.Annotation.Distance[i] / l.Annotation.Speed[i]   // not include turn weight
+			}
+		}
+		if m.LiveTrafficQuerier != nil && enableLiveTraffic {
+			if l.Annotation.BlockIncident[i] {
+				glog.Warningf("way %d on leg blocked by incident, set duration to infinity", wayID)
+				l.Duration = math.Inf(0)
+				l.Weight = math.Inf(0)
+				return nil
+			}
+			if trafficproxy.TrafficLevel(l.Annotation.LiveTrafficLevel[i]) == trafficproxy.TrafficLevel_CLOSED {
+				glog.Warningf("way %d on leg blocked by CLOSED flow, set duration to infinity", wayID)
+				l.Duration = math.Inf(0)
+				l.Weight = math.Inf(0)
+				return nil
+			}
+
+			if !trafficapplyingmodel.IsInvalidSpeed(l.Annotation.LiveTrafficSpeed[i]) { // valid live traffic speed
+				l.Annotation.DataSources[i] = liveTrafficSourceNameIndex
+				l.Annotation.Speed[i] = l.Annotation.LiveTrafficSpeed[i]
+				l.Annotation.Duration[i] = l.Annotation.Distance[i] / l.Annotation.Speed[i] // not include turn duration
+				l.Annotation.Weight[i] = l.Annotation.Distance[i] / l.Annotation.Speed[i]   // not include turn weight
+			}
+		}
+
+		newLegDuration += l.Annotation.Duration[i]
+		newLegWeight += l.Annotation.Weight[i]
+		if l.Annotation.DataSources[i] == liveTrafficSourceNameIndex {
+			appliedLiveTrafficSpeedCount++
+		}
+		if l.Annotation.DataSources[i] == historicalSpeedSourceNameIndex {
+			appliedHistoricalSpeedCount++
+		}
+	}
+	// metioned in doc, duration and weight in annotation does not include any on turn, so we calculate turn duration/weight by minus.
+	// https://github.com/Telenav/osrm-backend/blob/master-telenav/docs/http.md#annotation-object
+	legTurnDuration := l.Duration - sumOriginalAnnotationDuration
+	legTurnWeight := l.Weight - sumOriginalAnnotationWeight
```

Comment:
Keep original logic for how to calculate turns works, but might need pay attention for adjustment for the future.

Reply:
Sure. OSRM doesn't expose detailed turn duration/weights, we may have to figure out how to do it if we want to adjust them.